src/_c_internal_utils.c

```diff
@@ -1,11 +1,65 @@
 #define PY_SSIZE_T_CLEAN
 #include <Python.h>
+#ifdef __linux__
+#include <dlfcn.h>
+#endif
 #ifdef _WIN32
 #include <Objbase.h>
 #include <Shobjidl.h>
 #include <Windows.h>
 #endif
 
+static PyObject* mpl_display_is_valid(PyObject* module)
+{
+#ifdef __linux__
+    void* libX11;
+    // The getenv check is redundant but helps performance as it is much faster
+    // than dlopen().
+    if (getenv("DISPLAY")
+        && (libX11 = dlopen("libX11.so", RTLD_LAZY))) {
+        struct Display* display = NULL;
+        struct Display* (* XOpenDisplay)(char const*) =
+            dlsym(libX11, "XOpenDisplay");
+        int (* XCloseDisplay)(struct Display*) =
+            dlsym(libX11, "XCloseDisplay");
+        if (XOpenDisplay && XCloseDisplay
+                && (display = XOpenDisplay(NULL))) {
+            XCloseDisplay(display);
+        }
+        if (dlclose(libX11)) {
+            PyErr_SetString(PyExc_RuntimeError, dlerror());
+            return NULL;
+        }
+        if (display) {
+            Py_RETURN_TRUE;
+        }
+    }
+    void* libwayland_client;
+    if (getenv("WAYLAND_DISPLAY")
+        && (libwayland_client = dlopen("libwayland-client.so", RTLD_LAZY))) {
```

Comment:
Same here, which would have to check:
```suggestion
        && (libwayland_client = dlopen("libwayland-client.so.0", RTLD_LAZY))) {
```